File: app/src/main/java/djordjeh/architecture/mvp/ui/task/TaskPresenter.java
package djordjeh.architecture.mvp.ui.task;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import javax.inject.Inject;

import djordjeh.architecture.mvp.data.model.Task;
import djordjeh.architecture.mvp.data.source.TaskDataSource;
import djordjeh.architecture.mvp.ui.BasePresenterImpl;
import djordjeh.architecture.mvp.ui.SchedulersFacade;

public class TaskPresenter extends BasePresenterImpl<ContractTask.View> implements ContractTask.Presenter {

    private final TaskDataSource taskDataSource;
    private final SchedulersFacade schedulersFacade;

    @Inject
    TaskPresenter(@NonNull ContractTask.View view, TaskDataSource taskDataSource, SchedulersFacade schedulersFacade) {
        super(view);
        this.taskDataSource = taskDataSource;
        this.schedulersFacade = schedulersFacade;
    }

    @Override
    public void getTask(long taskId) {
        if (taskId != -1) {
            addDisposable(taskDataSource.task(taskId)
                    .subscribeOn(schedulersFacade.io())
                    .observeOn(schedulersFacade.mainThread())
                    .subscribe(view::showTask, view::showError)
            );
        } else view.showTask(new Task());
    }

    @Override
    public void saveTask(Task task) {
        if (isEmpty(task.getTitle())) {
            view.showEmptyTitleError(true);
        } else {
            view.showEmptyTitleError(false);
            addDisposable(taskDataSource.save(task)
                    .subscribeOn(schedulersFacade.io())
                    .observeOn(schedulersFacade.mainThread())
                    .subscribe(view::onTaskSaved, view::showError));
        }
    }

    private static boolean isEmpty(@Nullable String str) {
        return str == null || str.isEmpty();
    }
}

File: app/src/main/java/djordjeh/architecture/mvp/ui/task/TaskPresenter.kt
package djordjeh.architecture.mvp.ui.task

import javax.inject.Inject

import djordjeh.architecture.mvp.data.model.Task
import djordjeh.architecture.mvp.data.source.TaskDataSource
import djordjeh.architecture.mvp.ui.BasePresenterImpl
import djordjeh.architecture.mvp.ui.SchedulersFacade

class TaskPresenter @Inject
internal constructor(
        view: ContractTask.View,
        private val taskDataSource: TaskDataSource,
        private val schedulersFacade: SchedulersFacade) : BasePresenterImpl<ContractTask.View>(view), ContractTask.Presenter {

    override fun getTask(taskId: Long) {
        if (taskId > -1) {
            addDisposable(taskDataSource.task(taskId)
                    .subscribeOn(schedulersFacade.io())
                    .observeOn(schedulersFacade.mainThread())
                    .subscribe({ view.showTask(it) }, { view.showError(it) })
            )
        } else
            view.showTask(Task())
    }

    override fun saveTask(task: Task) {
        if (isEmpty(task.title)) {
            view.showEmptyTitleError(true)
        } else {
            view.showEmptyTitleError(false)
            addDisposable(taskDataSource.save(task)
                    .subscribeOn(schedulersFacade.io())
                    .observeOn(schedulersFacade.mainThread())
                    .subscribe({ view.onTaskSaved(it) }, { view.showError(it) }))
        }
    }

    private fun isEmpty(str: String?) = str?.isEmpty() ?: true
}
